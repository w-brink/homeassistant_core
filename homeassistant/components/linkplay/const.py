"""LinkPlay constants."""

from homeassistant.const import Platform

DOMAIN = "linkplay"
PLATFORMS = [Platform.MEDIA_PLAYER]
