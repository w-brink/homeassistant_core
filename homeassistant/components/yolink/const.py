"""Constants for the yolink integration."""

DOMAIN = "yolink"
MANUFACTURER = "YoLink"
ATTR_DEVICE = "devices"
ATTR_DEVICE_TYPE = "type"
ATTR_DEVICE_NAME = "name"
ATTR_DEVICE_STATE = "state"
ATTR_DEVICE_ID = "deviceId"
ATTR_TARGET_DEVICE = "target_device"
ATTR_VOLUME = "volume"
ATTR_TEXT_MESSAGE = "message"
ATTR_REPEAT = "repeat"
ATTR_TONE = "tone"
YOLINK_EVENT = f"{DOMAIN}_event"
YOLINK_OFFLINE_TIME = 32400

DEV_MODEL_WATER_METER_YS5007 = "YS5007"
DEV_MODEL_MULTI_OUTLET_YS6801 = "YS6801"
DEV_MODEL_TH_SENSOR_YS8004_UC = "YS8004-UC"
DEV_MODEL_TH_SENSOR_YS8004_EC = "YS8004-EC"
DEV_MODEL_TH_SENSOR_YS8014_UC = "YS8014-UC"
DEV_MODEL_TH_SENSOR_YS8014_EC = "YS8014-EC"
DEV_MODEL_TH_SENSOR_YS8017_UC = "YS8017-UC"
DEV_MODEL_TH_SENSOR_YS8017_EC = "YS8017-EC"
