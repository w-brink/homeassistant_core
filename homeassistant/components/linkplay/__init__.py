"""Support for LinkPlay devices."""

from linkplay.bridge import LinkPlayBridge
from linkplay.discovery import linkplay_factory_bridge

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_HOST
from homeassistant.core import HomeAssistant
from homeassistant.exceptions import ConfigEntryNotReady
from homeassistant.helpers.aiohttp_client import async_get_clientsession

from .const import PLATFORMS


class LinkPlayData:
    """Data for LinkPlay."""

    bridge: LinkPlayBridge


type LinkPlayConfigEntry = ConfigEntry[LinkPlayData]


async def async_setup_entry(hass: HomeAssistant, entry: LinkPlayConfigEntry) -> bool:
    """Async setup hass config entry. Called when an entry has been setup."""

    session = async_get_clientsession(hass)
    if (
        bridge := await linkplay_factory_bridge(entry.data[CONF_HOST], session)
    ) is None:
        raise ConfigEntryNotReady(
            f"Failed to connect to LinkPlay device at {entry.data[CONF_HOST]}"
        )

    entry.runtime_data = LinkPlayData()
    entry.runtime_data.bridge = bridge
    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
    return True


async def async_unload_entry(hass: HomeAssistant, entry: LinkPlayConfigEntry) -> bool:
    """Unload a config entry."""

    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
