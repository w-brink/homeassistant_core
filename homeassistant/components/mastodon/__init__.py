"""The Mastodon integration."""

from __future__ import annotations

from mastodon.Mastodon import Mastodon, MastodonError

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import (
    CONF_ACCESS_TOKEN,
    CONF_CLIENT_ID,
    CONF_CLIENT_SECRET,
    CONF_NAME,
    Platform,
)
from homeassistant.core import HomeAssistant
from homeassistant.exceptions import ConfigEntryNotReady
from homeassistant.helpers import discovery

from .const import CONF_BASE_URL, DOMAIN
from .utils import create_mastodon_client


async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Set up Mastodon from a config entry."""

    try:
        client, _, _ = await hass.async_add_executor_job(
            setup_mastodon,
            entry,
        )

    except MastodonError as ex:
        raise ConfigEntryNotReady("Failed to connect") from ex

    assert entry.unique_id

    await discovery.async_load_platform(
        hass,
        Platform.NOTIFY,
        DOMAIN,
        {CONF_NAME: entry.title, "client": client},
        {},
    )

    return True


def setup_mastodon(entry: ConfigEntry) -> tuple[Mastodon, dict, dict]:
    """Get mastodon details."""
    client = create_mastodon_client(
        entry.data[CONF_BASE_URL],
        entry.data[CONF_CLIENT_ID],
        entry.data[CONF_CLIENT_SECRET],
        entry.data[CONF_ACCESS_TOKEN],
    )

    instance = client.instance()
    account = client.account_verify_credentials()

    return client, instance, account
