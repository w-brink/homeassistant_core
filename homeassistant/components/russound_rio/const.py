"""Constants used for Russound RIO."""

import asyncio

from aiorussound import CommandException
from aiorussound.const import FeatureFlag

from homeassistant.components.media_player import MediaPlayerEntityFeature

DOMAIN = "russound_rio"

RUSSOUND_RIO_EXCEPTIONS = (
    CommandException,
    ConnectionRefusedError,
    TimeoutError,
    asyncio.CancelledError,
)


class NoPrimaryControllerException(Exception):
    """Thrown when the Russound device is not the primary unit in the RNET stack."""


CONNECT_TIMEOUT = 5

MP_FEATURES_BY_FLAG = {
    FeatureFlag.COMMANDS_ZONE_MUTE_OFF_ON: MediaPlayerEntityFeature.VOLUME_MUTE
}
