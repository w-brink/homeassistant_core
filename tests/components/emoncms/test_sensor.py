"""Test emoncms sensor."""

from typing import Any
from unittest.mock import AsyncMock

from freezegun.api import FrozenDateTimeFactory
import pytest
from syrupy.assertion import SnapshotAssertion

from homeassistant.components.emoncms.const import CONF_ONLY_INCLUDE_FEEDID, DOMAIN
from homeassistant.components.sensor import DOMAIN as SENSOR_DOMAIN
from homeassistant.const import CONF_API_KEY, CONF_ID, CONF_PLATFORM, CONF_URL
from homeassistant.core import HomeAssistant
from homeassistant.helpers.typing import ConfigType
from homeassistant.setup import async_setup_component

from .conftest import EMONCMS_FAILURE, FEEDS, get_feed

from tests.common import async_fire_time_changed

YAML = {
    CONF_PLATFORM: "emoncms",
    CONF_API_KEY: "my_api_key",
    CONF_ID: 1,
    CONF_URL: "http://1.1.1.1",
    CONF_ONLY_INCLUDE_FEEDID: [1, 2],
    "scan_interval": 30,
}


@pytest.fixture
def emoncms_yaml_config() -> ConfigType:
    """Mock emoncms configuration from yaml."""
    return {"sensor": YAML}


def get_entity_ids(feeds: list[dict[str, Any]]) -> list[str]:
    """Get emoncms entity ids."""
    return [
        f"{SENSOR_DOMAIN}.{DOMAIN}_{feed["name"].replace(' ', '_')}" for feed in feeds
    ]


def get_feeds(nbs: list[int]) -> list[dict[str, Any]]:
    """Get feeds."""
    return [feed for feed in FEEDS if feed["id"] in str(nbs)]


async def test_coordinator_update(
    hass: HomeAssistant,
    emoncms_yaml_config: ConfigType,
    snapshot: SnapshotAssertion,
    emoncms_client: AsyncMock,
    caplog: pytest.LogCaptureFixture,
    freezer: FrozenDateTimeFactory,
) -> None:
    """Test coordinator update."""
    emoncms_client.async_request.return_value = {
        "success": True,
        "message": [get_feed(1, unit="°C")],
    }
    await async_setup_component(hass, SENSOR_DOMAIN, emoncms_yaml_config)
    await hass.async_block_till_done()
    feeds = get_feeds([1])
    for entity_id in get_entity_ids(feeds):
        state = hass.states.get(entity_id)
        assert state == snapshot(name=entity_id)

    async def skip_time() -> None:
        freezer.tick(60)
        async_fire_time_changed(hass)
        await hass.async_block_till_done(wait_background_tasks=True)

    emoncms_client.async_request.return_value = {
        "success": True,
        "message": [get_feed(1, unit="°C", value=24.04, timestamp=1665509670)],
    }

    await skip_time()

    for entity_id in get_entity_ids(feeds):
        state = hass.states.get(entity_id)
        assert state.attributes["LastUpdated"] == 1665509670
        assert state.state == "24.04"

    emoncms_client.async_request.return_value = EMONCMS_FAILURE

    await skip_time()

    assert f"Error fetching {DOMAIN}_coordinator data" in caplog.text
