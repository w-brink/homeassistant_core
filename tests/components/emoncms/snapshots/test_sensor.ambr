# serializer version: 1
# name: test_coordinator_update[sensor.emoncms_parameter_1]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'FeedId': '1',
      'FeedName': 'parameter 1',
      'LastUpdated': 1665509570,
      'LastUpdatedStr': '2022-10-11T10:32:50-07:00',
      'Size': '35809224',
      'Tag': 'tag',
      'UserId': '1',
      'device_class': 'temperature',
      'friendly_name': 'EmonCMS parameter 1',
      'state_class': <SensorStateClass.MEASUREMENT: 'measurement'>,
      'unit_of_measurement': <UnitOfTemperature.CELSIUS: 'Â°C'>,
    }),
    'context': <ANY>,
    'entity_id': 'sensor.emoncms_parameter_1',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '18.04',
  })
# ---
